jupyterhub/tests/test_scopes.py

```diff
@@ -0,0 +1,195 @@
+        return True
+
+    @needs_scope('services')
+    def service_thing(self, service_name):
+        return True
+
+    @needs_scope('users')
+    def other_thing(self, other_name):
+        return True
+
+
+@mark.parametrize(
+    "scopes, method, arguments, is_allowed",
+    [
+        (['users'], 'user_thing', ('user',), True),
+        (['users'], 'user_thing', ('michael',), True),
+        ([''], 'user_thing', ('michael',), False),
+        (['read:users'], 'user_thing', ('gob',), False),
+        (['read:users'], 'user_thing', ('michael',), False),
+        (['users!user=george'], 'user_thing', ('george',), True),
+        (['users!user=george'], 'user_thing', ('oscar',), False),
+        (['users!user=george', 'users!user=oscar'], 'user_thing', ('oscar',), True),
+        (['users:servers'], 'server_thing', ('user1', 'server_1'), True),
+        (['users:servers'], 'server_thing', ('user1', ''), True),
+        (['users:servers'], 'server_thing', ('user1', None), True),
+        (
+            ['users:servers!server=maeby/bluth'],
+            'server_thing',
+            ('maeby', 'bluth'),
+            True,
+        ),
+        (['users:servers!server=maeby/bluth'], 'server_thing', ('gob', 'bluth'), False),
+        (
+            ['users:servers!server=maeby/bluth'],
+            'server_thing',
+            ('maybe', 'bluth2'),
+            False,
+        ),
+        (['services'], 'service_thing', ('service1',), True),
+        (
+            ['users!user=george', 'read:groups!group=bluths'],
+            'group_thing',
+            ('bluths',),
+            True,
+        ),
+        (
+            ['users!user=george', 'read:groups!group=bluths'],
+            'group_thing',
+            ('george',),
+            False,
+        ),
+        (
+            ['groups!group=george', 'read:groups!group=bluths'],
+            'group_thing',
+            ('george',),
+            False,
+        ),
+        (['users'], 'other_thing', ('gob',), True),
+        (['read:users'], 'other_thing', ('gob',), False),
+        (['users!user=gob'], 'other_thing', ('gob',), False),
+        (['users!user=gob'], 'other_thing', ('maeby',), False),
+    ],
+)
+def test_scope_method_access(scopes, method, arguments, is_allowed):
+    obj = MockAPIHandler()
+    obj.scopes = scopes
+    api_call = getattr(obj, method)
+    if is_allowed:
+        assert api_call(*arguments)
+    else:
+        with pytest.raises(web.HTTPError):
+            api_call(*arguments)
+
+
+@mark.parametrize(
+    "name, status_code",
+    [
+        ('martha', 200),
+        ('michael', 200),
+        ('gob', 200),
+        ('tobias', 403),
+        ('ann', 404),  # this leaks a bit of information, what do we think about that?
+    ],
+)
+async def test_expand_groups(app, name, status_code):
+    app.init_db()
+    group = orm.Group(name='bluth')
+    app.db.add(group)
+    for user_name in ['gob', 'michael']:
+        user = orm.User(name=user_name)
```

Comment:
Try using `utils.add_user` to create the users in this (and other) tests. It might be related to the db failures, since this could end up creating multiple users with the same name (I would expect that to fail a uniqueness constraint, but still best to use the utility).